src/main/java/net/dv8tion/jda/api/entities/Message.java

```diff
@@ -1114,113 +1106,192 @@ public String getFileName()
-         *         If this is not an image attachment
-         *
-         * @return {@link net.dv8tion.jda.api.entities.Icon Icon} for this image attachment
-         *
-         * @since  3.4.0
+         * @return {@link java.util.concurrent.CompletableFuture} - Type: {@link java.io.InputStream}
          */
-        public Icon getAsIcon() throws IOException
+        public CompletableFuture<InputStream> retrieveData() // it is expected that the response is closed by the callback!
         {
-            if (!isImage())
-                throw new IllegalStateException("Cannot create an Icon out of this attachment. This is not an image.");
-            AtomicReference<Icon> icon = new AtomicReference<>();
-            withInputStream((in) -> icon.set(Icon.from(in)));
-            return icon.get();
+            CompletableFuture<InputStream> future = new CompletableFuture<>();
+            Request req = getRequest();
+            OkHttpClient httpClient = getJDA().getHttpClient();
+            httpClient.newCall(req).enqueue(new Callback()
+            {
+                @Override
+                public void onFailure(Call call, IOException e)
+                {
+                    future.completeExceptionally(new UncheckedIOException(e));
+                }
+
+                @Override
+                public void onResponse(Call call, Response response) throws IOException
+                {
+                    if (response.isSuccessful())
+                    {
+                        InputStream body = Requester.getBody(response);
+                        if (!future.complete(body))
+                            IOUtil.silentClose(response);
+                    }
+                    else
+                    {
+                        future.completeExceptionally(new HttpException(response.code() + ": " + response.message()));
+                        IOUtil.silentClose(response);
+                    }
+                }
+            });
+            return future;
         }
 
         /**
-         * Downloads this attachment to given File
+         * Downloads the attachment to a file with the {@link #getFileName()} into the current working directory.
+         * <br>This will download the file using the {@link net.dv8tion.jda.api.JDA#getCallbackPool() callback pool}.
+         * Alternatively you can use {@link #retrieveData()} and use a continuation with a different executor.
          *
-         * @param  file
-         *         The file, where the attachment will get downloaded to
+         * <h2>Example</h2>
+         * <pre>{@code
+         * public void saveLocally(Message.Attachment attachment)
+         * {
+         *     attachment.downloadToFile()
+         *         .thenAccept(file -> System.out.println("Saved attachment to " + file.getName()))
+         *         .exceptionally(t ->
+         *     { // handle failure
+         *         t.printStackTrace();
+         *         return null;
```

Comment:
Indention of the example's `exceptionally` callback is bork.

Reply:
This indentation is intentional, though

Reply:
Does it render differently than it appears in the jdoc?
Jdoc source has the callback's code inline with the method chain. It should be 1 level deeper.